<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="fr">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Data analysis &#8212; test</title>
    
    <link rel="stylesheet" href="_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/bootswatch-3.3.6/yeti/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.10.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  false,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/translations.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="_static/bootstrap-3.3.6/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="_static/bootstrap-sphinx.js"></script>
    <link rel="shortcut icon" href="_static/canopee_ico.ico"/>
    <link rel="search" title="Recherche" href="search.html" />
    <link rel="next" title="Data visualisation" href="data-visualisation.html" />
    <link rel="prev" title="Keeping track of your work" href="keeping-track-of-your-work.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body>

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="index.html">
          test</a>
        <span class="navbar-text navbar-version pull-left"><b></b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="index.html">Yo!test <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="preface.html">Preface</a><ul>
<li class="toctree-l2"><a class="reference internal" href="preface.html#who-should-read-this-book">Who should read this book</a></li>
<li class="toctree-l2"><a class="reference internal" href="preface.html#licence">Licence</a></li>
<li class="toctree-l2"><a class="reference internal" href="preface.html#source-code">Source code</a></li>
<li class="toctree-l2"><a class="reference internal" href="preface.html#feedback">Feedback</a></li>
<li class="toctree-l2"><a class="reference internal" href="preface.html#acknowledgements">Acknowledgements</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="getting-some-motivation.html">Getting some motivation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="getting-some-motivation.html#why-you-should-care-about-computing-and-coding">Why you should care about computing and coding</a></li>
<li class="toctree-l2"><a class="reference internal" href="getting-some-motivation.html#what-you-can-expect-to-learn-here">What you can expect to learn here</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="how-to-think-like-a-computer.html">How to think like a computer</a><ul>
<li class="toctree-l2"><a class="reference internal" href="how-to-think-like-a-computer.html#binary-bits-and-bytes">Binary, bits and bytes</a></li>
<li class="toctree-l2"><a class="reference internal" href="how-to-think-like-a-computer.html#character-encodings">Character encodings</a></li>
<li class="toctree-l2"><a class="reference internal" href="how-to-think-like-a-computer.html#real-numbers">Real numbers</a></li>
<li class="toctree-l2"><a class="reference internal" href="how-to-think-like-a-computer.html#boolean-logic">Boolean logic</a></li>
<li class="toctree-l2"><a class="reference internal" href="how-to-think-like-a-computer.html#the-microprocessor">The microprocessor</a></li>
<li class="toctree-l2"><a class="reference internal" href="how-to-think-like-a-computer.html#computer-memory">Computer memory</a></li>
<li class="toctree-l2"><a class="reference internal" href="how-to-think-like-a-computer.html#key-concepts">Key concepts</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="first-steps-towards-automation.html">First steps towards automation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="first-steps-towards-automation.html#first-things-first-how-to-find-help">First things first, how to find help</a></li>
<li class="toctree-l2"><a class="reference internal" href="first-steps-towards-automation.html#creating-a-new-directory-for-our-project">Creating a new directory for our project</a></li>
<li class="toctree-l2"><a class="reference internal" href="first-steps-towards-automation.html#downloading-the-swiss-prot-knowledge-base">Downloading the Swiss-Prot knowledge base</a></li>
<li class="toctree-l2"><a class="reference internal" href="first-steps-towards-automation.html#creating-a-work-flow-using-pipes">Creating a work flow using pipes</a></li>
<li class="toctree-l2"><a class="reference internal" href="first-steps-towards-automation.html#examining-files-without-modifying-them">Examining files, without modifying them</a></li>
<li class="toctree-l2"><a class="reference internal" href="first-steps-towards-automation.html#finding-fasta-identifier-lines-corresponding-to-human-proteins">Finding FASTA identifier lines corresponding to human proteins</a></li>
<li class="toctree-l2"><a class="reference internal" href="first-steps-towards-automation.html#extracting-the-uniprot-identifiers">Extracting the UniProt identifiers</a></li>
<li class="toctree-l2"><a class="reference internal" href="first-steps-towards-automation.html#using-redirection-to-create-an-output-file">Using redirection to create an output file</a></li>
<li class="toctree-l2"><a class="reference internal" href="first-steps-towards-automation.html#viewing-the-command-history">Viewing the command history</a></li>
<li class="toctree-l2"><a class="reference internal" href="first-steps-towards-automation.html#clearing-the-terminal-window">Clearing the terminal window</a></li>
<li class="toctree-l2"><a class="reference internal" href="first-steps-towards-automation.html#copying-and-renaming-files">Copying and renaming files</a></li>
<li class="toctree-l2"><a class="reference internal" href="first-steps-towards-automation.html#removing-files-and-directories">Removing files and directories</a></li>
<li class="toctree-l2"><a class="reference internal" href="first-steps-towards-automation.html#key-concepts">Key concepts</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="structuring-and-storing-data.html">Structuring and storing data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="structuring-and-storing-data.html#data-structures">Data structures</a></li>
<li class="toctree-l2"><a class="reference internal" href="structuring-and-storing-data.html#data-persistence">Data persistence</a></li>
<li class="toctree-l2"><a class="reference internal" href="structuring-and-storing-data.html#the-beauty-of-plain-text-files">The beauty of plain text files</a></li>
<li class="toctree-l2"><a class="reference internal" href="structuring-and-storing-data.html#useful-plain-text-file-formats">Useful plain text file formats</a></li>
<li class="toctree-l2"><a class="reference internal" href="structuring-and-storing-data.html#tidy-data">Tidy data</a></li>
<li class="toctree-l2"><a class="reference internal" href="structuring-and-storing-data.html#find-a-good-text-editor-and-learn-how-to-use-it">Find a good text editor and learn how to use it</a></li>
<li class="toctree-l2"><a class="reference internal" href="structuring-and-storing-data.html#key-concepts">Key concepts</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="keeping-track-of-your-work.html">Keeping track of your work</a><ul>
<li class="toctree-l2"><a class="reference internal" href="keeping-track-of-your-work.html#what-is-git">What is Git?</a></li>
<li class="toctree-l2"><a class="reference internal" href="keeping-track-of-your-work.html#first-time-configuration-of-git">First time configuration of Git</a></li>
<li class="toctree-l2"><a class="reference internal" href="keeping-track-of-your-work.html#initialise-the-project">Initialise the project</a></li>
<li class="toctree-l2"><a class="reference internal" href="keeping-track-of-your-work.html#create-a-script-for-downloading-the-swissprot-fasta-file">Create a script for downloading the SwissProt FASTA file</a></li>
<li class="toctree-l2"><a class="reference internal" href="keeping-track-of-your-work.html#improve-script-for-downloading-swissprot-fasta-file">Improve script for downloading SwissProt FASTA file</a></li>
<li class="toctree-l2"><a class="reference internal" href="keeping-track-of-your-work.html#create-script-for-counting-the-number-of-proteins-in-a-genome">Create script for counting the number of proteins in a genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="keeping-track-of-your-work.html#more-useful-git-commands">More useful git commands</a></li>
<li class="toctree-l2"><a class="reference internal" href="keeping-track-of-your-work.html#key-concepts">Key concepts</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Data analysis</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#what-is-python">What is Python?</a></li>
<li class="toctree-l2"><a class="reference internal" href="#using-python-in-interactive-mode">Using Python in interactive mode</a></li>
<li class="toctree-l2"><a class="reference internal" href="#variables">Variables</a></li>
<li class="toctree-l2"><a class="reference internal" href="#determining-the-gc-count-of-a-sequence">Determining the GC count of a sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="#creating-reusable-functions">Creating reusable functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="#list-slicing">List slicing</a></li>
<li class="toctree-l2"><a class="reference internal" href="#loops">Loops</a></li>
<li class="toctree-l2"><a class="reference internal" href="#creating-a-sliding-window-gc-content-function">Creating a sliding window GC-content function</a></li>
<li class="toctree-l2"><a class="reference internal" href="#downloading-the-genome">Downloading the genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="#reading-and-writing-files">Reading and writing files</a></li>
<li class="toctree-l2"><a class="reference internal" href="#creating-a-function-for-reading-in-the-streptomyces-sequence">Creating a function for reading in the <em>Streptomyces</em> sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="#writing-out-the-sliding-window-analysis">Writing out the sliding window analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="#key-concepts">Key concepts</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="data-visualisation.html">Data visualisation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="data-visualisation.html#starting-r-and-loading-the-iris-flower-data-set">Starting R and loading the Iris flower data set</a></li>
<li class="toctree-l2"><a class="reference internal" href="data-visualisation.html#understanding-the-structure-of-the-iris-data-set">Understanding the structure of the <code class="docutils literal"><span class="pre">iris</span></code> data set</a></li>
<li class="toctree-l2"><a class="reference internal" href="data-visualisation.html#a-note-on-statistics-in-r">A note on statistics in R</a></li>
<li class="toctree-l2"><a class="reference internal" href="data-visualisation.html#default-plotting-in-r">Default plotting in R</a></li>
<li class="toctree-l2"><a class="reference internal" href="data-visualisation.html#installing-the-ggplot2-package">Installing the ggplot2 package</a></li>
<li class="toctree-l2"><a class="reference internal" href="data-visualisation.html#loading-the-ggplot2-package">Loading the ggplot2 package</a></li>
<li class="toctree-l2"><a class="reference internal" href="data-visualisation.html#plotting-using-ggplot2">Plotting using ggplot2</a></li>
<li class="toctree-l2"><a class="reference internal" href="data-visualisation.html#available-geoms">Available «&nbsp;Geoms&nbsp;»</a></li>
<li class="toctree-l2"><a class="reference internal" href="data-visualisation.html#scripting-data-visualisation">Scripting data visualisation</a></li>
<li class="toctree-l2"><a class="reference internal" href="data-visualisation.html#faceting">Faceting</a></li>
<li class="toctree-l2"><a class="reference internal" href="data-visualisation.html#adding-more-colour">Adding more colour</a></li>
<li class="toctree-l2"><a class="reference internal" href="data-visualisation.html#purpose-of-data-visualisation">Purpose of data visualisation</a></li>
<li class="toctree-l2"><a class="reference internal" href="data-visualisation.html#conveying-a-message-to-an-audience">Conveying a message to an audience</a></li>
<li class="toctree-l2"><a class="reference internal" href="data-visualisation.html#writing-a-caption">Writing a caption</a></li>
<li class="toctree-l2"><a class="reference internal" href="data-visualisation.html#other-useful-tools-for-scripting-the-generation-of-figures">Other useful tools for scripting the generation of figures</a></li>
<li class="toctree-l2"><a class="reference internal" href="data-visualisation.html#key-concepts">Key concepts</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="collaborating-on-projects.html">Collaborating on projects</a><ul>
<li class="toctree-l2"><a class="reference internal" href="collaborating-on-projects.html#find-a-friend">Find a friend</a></li>
<li class="toctree-l2"><a class="reference internal" href="collaborating-on-projects.html#get-a-github-or-bitbucket-account">Get a GitHub or BitBucket account</a></li>
<li class="toctree-l2"><a class="reference internal" href="collaborating-on-projects.html#pushing-the-project-to-a-remote-server">Pushing the project to a remote server</a></li>
<li class="toctree-l2"><a class="reference internal" href="collaborating-on-projects.html#collaboration-using-git">Collaboration using Git</a></li>
<li class="toctree-l2"><a class="reference internal" href="collaborating-on-projects.html#working-in-parallel">Working in parallel</a></li>
<li class="toctree-l2"><a class="reference internal" href="collaborating-on-projects.html#resolving-conflicts">Resolving conflicts</a></li>
<li class="toctree-l2"><a class="reference internal" href="collaborating-on-projects.html#key-concepts">Key concepts</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="creating-scientific-documents.html">Creating scientific documents</a><ul>
<li class="toctree-l2"><a class="reference internal" href="creating-scientific-documents.html#starting-the-manuscript">Starting the manuscript</a></li>
<li class="toctree-l2"><a class="reference internal" href="creating-scientific-documents.html#converting-the-document-to-html-using-pandoc">Converting the document to HTML using pandoc</a></li>
<li class="toctree-l2"><a class="reference internal" href="creating-scientific-documents.html#adding-a-figure">Adding a figure</a></li>
<li class="toctree-l2"><a class="reference internal" href="creating-scientific-documents.html#converting-the-document-to-pdf">Converting the document to PDF</a></li>
<li class="toctree-l2"><a class="reference internal" href="creating-scientific-documents.html#reference-management">Reference management</a></li>
<li class="toctree-l2"><a class="reference internal" href="creating-scientific-documents.html#adding-meta-data">Adding meta data</a></li>
<li class="toctree-l2"><a class="reference internal" href="creating-scientific-documents.html#benefits-of-using-pandoc-and-plain-text-files">Benefits of using Pandoc and plain text files</a></li>
<li class="toctree-l2"><a class="reference internal" href="creating-scientific-documents.html#alternative-plain-text-approaches">Alternative plain text approaches</a></li>
<li class="toctree-l2"><a class="reference internal" href="creating-scientific-documents.html#using-non-plain-text-file-formats">Using non-plain text file formats</a></li>
<li class="toctree-l2"><a class="reference internal" href="creating-scientific-documents.html#key-concepts">Key concepts</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="automation-is-your-friend.html">Automation is your friend</a><ul>
<li class="toctree-l2"><a class="reference internal" href="automation-is-your-friend.html#introduction-to-make">Introduction to <code class="docutils literal"><span class="pre">make</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="automation-is-your-friend.html#creating-a-makefile">Creating a <code class="docutils literal"><span class="pre">Makefile</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="automation-is-your-friend.html#key-concepts">Key concepts</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="practical-problem-solving.html">Practical problem solving</a><ul>
<li class="toctree-l2"><a class="reference internal" href="practical-problem-solving.html#stop-and-think">Stop and think</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical-problem-solving.html#measuring-success">Measuring success</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical-problem-solving.html#more-string-processing">More string processing</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical-problem-solving.html#extracting-the-organism-name">Extracting the organism name</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical-problem-solving.html#only-running-tests-when-the-module-is-called-directly">Only running tests when the module is called directly</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical-problem-solving.html#counting-the-number-of-unique-organisms">Counting the number of unique organisms</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical-problem-solving.html#finding-the-number-of-variants-per-species-and-the-number-of-proteins-per-variant">Finding the number of variants per species and the number of proteins per variant</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical-problem-solving.html#key-concepts">Key concepts</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="working-remotely.html">Working remotely</a><ul>
<li class="toctree-l2"><a class="reference internal" href="working-remotely.html#jargon-busting">Jargon busting</a></li>
<li class="toctree-l2"><a class="reference internal" href="working-remotely.html#logging-in-using-the-secure-shell-protocol">Logging in using the Secure Shell protocol</a></li>
<li class="toctree-l2"><a class="reference internal" href="working-remotely.html#copying-files-using-secure-copy">Copying files using Secure Copy</a></li>
<li class="toctree-l2"><a class="reference internal" href="working-remotely.html#password-less-authentication-using-ssh-keys">Password-less authentication using SSH keys</a></li>
<li class="toctree-l2"><a class="reference internal" href="working-remotely.html#managing-your-login-details-using-ssh-config">Managing your login details using SSH config</a></li>
<li class="toctree-l2"><a class="reference internal" href="working-remotely.html#executing-long-running-commands-on-remote-hosts">Executing long running commands on remote hosts</a></li>
<li class="toctree-l2"><a class="reference internal" href="working-remotely.html#key-concepts">Key concepts</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="managing-your-system.html">Managing your system</a><ul>
<li class="toctree-l2"><a class="reference internal" href="managing-your-system.html#the-file-system">The file system</a></li>
<li class="toctree-l2"><a class="reference internal" href="managing-your-system.html#the-path-environment-variable">The <code class="docutils literal"><span class="pre">PATH</span></code> environment variable</a></li>
<li class="toctree-l2"><a class="reference internal" href="managing-your-system.html#the-root-user">The <code class="docutils literal"><span class="pre">root</span></code> user</a></li>
<li class="toctree-l2"><a class="reference internal" href="managing-your-system.html#using-package-managers">Using package managers</a></li>
<li class="toctree-l2"><a class="reference internal" href="managing-your-system.html#compiling-software-from-source">Compiling software from source</a></li>
<li class="toctree-l2"><a class="reference internal" href="managing-your-system.html#installing-python-packages">Installing Python packages</a></li>
<li class="toctree-l2"><a class="reference internal" href="managing-your-system.html#installing-r-packages">Installing R packages</a></li>
<li class="toctree-l2"><a class="reference internal" href="managing-your-system.html#installing-perl-modules">Installing Perl modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="managing-your-system.html#installing-latex-packages">Installing Latex packages</a></li>
<li class="toctree-l2"><a class="reference internal" href="managing-your-system.html#automated-provisioning">Automated provisioning</a></li>
<li class="toctree-l2"><a class="reference internal" href="managing-your-system.html#key-concepts">Key concepts</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="next-steps.html">Next steps</a><ul>
<li class="toctree-l2"><a class="reference internal" href="next-steps.html#apply-what-you-have-learnt">Apply what you have learnt</a></li>
<li class="toctree-l2"><a class="reference internal" href="next-steps.html#be-social">Be social</a></li>
<li class="toctree-l2"><a class="reference internal" href="next-steps.html#recommended-reading">Recommended reading</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary</a></li>
</ul>
</ul>
</li>
              
            
            
            
            
            
              <li class="hidden-sm"></li>
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
      <div class="col-md-3">
        <div id="sidebar" class="bs-sidenav" role="complementary"><ul>
<li><a class="reference internal" href="#">Data analysis</a><ul>
<li><a class="reference internal" href="#what-is-python">What is Python?</a></li>
<li><a class="reference internal" href="#using-python-in-interactive-mode">Using Python in interactive mode</a></li>
<li><a class="reference internal" href="#variables">Variables</a></li>
<li><a class="reference internal" href="#determining-the-gc-count-of-a-sequence">Determining the GC count of a sequence</a></li>
<li><a class="reference internal" href="#creating-reusable-functions">Creating reusable functions</a></li>
<li><a class="reference internal" href="#list-slicing">List slicing</a></li>
<li><a class="reference internal" href="#loops">Loops</a></li>
<li><a class="reference internal" href="#creating-a-sliding-window-gc-content-function">Creating a sliding window GC-content function</a></li>
<li><a class="reference internal" href="#downloading-the-genome">Downloading the genome</a></li>
<li><a class="reference internal" href="#reading-and-writing-files">Reading and writing files</a></li>
<li><a class="reference internal" href="#creating-a-function-for-reading-in-the-streptomyces-sequence">Creating a function for reading in the <em>Streptomyces</em> sequence</a></li>
<li><a class="reference internal" href="#writing-out-the-sliding-window-analysis">Writing out the sliding window analysis</a></li>
<li><a class="reference internal" href="#key-concepts">Key concepts</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    <div class="col-md-9 content">
      
  <div class="section" id="data-analysis">
<h1>Data analysis</h1>
<p><em>Streptomyces coelicolor</em> is a soil-dwelling bacterium with a complex life-cycle
involving mycelial growth and sporulation. It is of particular interest in that
it can produce a range of natural products of pharmaceutical relevance. In fact
it is the major source of natural antibiotics.</p>
<p>In 2002 the genome of <em>Streptomyces coelicolor</em>  A3(2), the model actinomycete
organism, was sequenced. In this chapter we will do some basic bioinformatics
on this genome to illustrate some fundamental concepts of computer programming.</p>
<p>One feature of interest when examining DNA is the guanine-cytosine (GC) content.
DNA with high GC-content is more stable than DNA with low GC-content.
The GC-content is defined as:</p>
<div class="math">
\[100 *  \frac{G + C}{A + T + G + C}\]</div>
<p>To solve this problem we need to write code to be able to read in sequence data
from a file, process the input to calculate the local GC-content and write out
the results to another file. In the process the computational concepts of
variables, functions and loops will be introduced.</p>
<p>In this chapter we will use the Python scripting language to perform our data
analysis.</p>
<div class="section" id="what-is-python">
<h2>What is Python?</h2>
<p>Python is a high-level scripting language that is growing in popularity in the
scientific community. It uses a syntax that is relatively easy to get to grips
with and which encourages code readability.</p>
</div>
<div class="section" id="using-python-in-interactive-mode">
<h2>Using Python in interactive mode</h2>
<p>To start off with we will make use of Python using its
<a class="reference external" href="https://docs.python.org/2/tutorial/interpreter.html#interactive-mode">interactive mode</a>,
which means that we can type Python commands straight into the terminal. In fact
when working with Python in interactive mode one can think of it as switching
the terminal’s shell from Bash to Python.</p>
<p>To start Python in its interactive mode simply type <code class="docutils literal"><span class="pre">python</span></code> into the terminal.</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>$ python
Python 2.7.10 (default, Jul 14 2015, 19:46:27)
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.39)] on darwin
Type &quot;help&quot;, &quot;copyright&quot;, &quot;credits&quot; or &quot;license&quot; for more information.
&gt;&gt;&gt;
</pre></div>
</div>
<p>Note that this prints out information about the version of Python that is being used
and how it was compiled before leaving you at the interactive prompt. In this instance
I am using Python version 2.7.10.</p>
<p>The three greater than signs (<code class="docutils literal"><span class="pre">&gt;&gt;&gt;</span></code>) represent the primary prompt into which
commands can be entered.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="mi">1</span> <span class="o">+</span> <span class="mi">2</span>
<span class="go">3</span>
</pre></div>
</div>
<p>The secondary prompt is represented by three dots (<code class="docutils literal"><span class="pre">...</span></code>) and denotes a
continuation line.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">line</span> <span class="o">=</span> <span class="s2">&quot;&gt;sp|Q6GZX4|001R_FRG3G Putative transcription factor 001R&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;&gt;&quot;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">print</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">&gt;sp|Q6GZX4|001R_FRG3G Putative transcription factor 001R</span>
</pre></div>
</div>
</div>
<div class="section" id="variables">
<h2>Variables</h2>
<p>A variable is a means of storing a piece of information using using a
descriptive name. The use of variables is encouraged as it allows us to
avoid having to repeat ourselves, the <a class="reference internal" href="glossary.html#term-dry"><span class="xref std std-term">DRY</span></a> principle.</p>
<p>In Python variables are assigned using the equals sign.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pi</span> <span class="o">=</span> <span class="mf">3.14</span>
</pre></div>
</div>
<p>When naming variables being explicit is more important than being succinct.
One reason for this is that you will spend more time reading your code than
you will writing it. Avoiding the mental overhead of trying to understand
what all the acronyms mean is a good thing. For example, suppose that we
wanted to create a variable for storing the radius of a circle. Please
avoid the temptation of naming the variable <code class="docutils literal"><span class="pre">r</span></code>, and go for the longer
but more explicit name <code class="docutils literal"><span class="pre">radius</span></code>.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">radius</span> <span class="o">=</span> <span class="mf">1.5</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Many modern text editors have auto complete functionality so
longer variable names does not necessarily need to mean that
there is more typing required. Remember, spend time
finding a text editor that works well for you!</p>
</div>
</div>
<div class="section" id="determining-the-gc-count-of-a-sequence">
<h2>Determining the GC count of a sequence</h2>
<p>Suppose that we had a <a class="reference internal" href="glossary.html#term-string"><span class="xref std std-term">string</span></a> representing a DNA sequence.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dna_string</span> <span class="o">=</span> <span class="s2">&quot;attagcgcaatctaactacactactgccgcgcggcatatatttaaatata&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">dna_string</span><span class="p">)</span>
<span class="go">attagcgcaatctaactacactactgccgcgcggcatatatttaaatata</span>
</pre></div>
</div>
<p>A string is a data type for representing text. As such it is not ideal for data
processing purposes. In this case the DNA sequence would be better represented
using a <a class="reference internal" href="glossary.html#term-list"><span class="xref std std-term">list</span></a>, with each item in the list representing a DNA letter.</p>
<p>In Python we can convert a string into a list using the built-in <code class="xref py py-func docutils literal"><span class="pre">list()</span></code>
function.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dna_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">dna_string</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">dna_list</span><span class="p">)</span>  
<span class="go">[&#39;a&#39;, &#39;t&#39;, &#39;t&#39;, &#39;a&#39;, &#39;g&#39;, &#39;c&#39;, &#39;g&#39;, &#39;c&#39;, &#39;a&#39;, &#39;a&#39;, &#39;t&#39;, &#39;c&#39;, &#39;t&#39;, &#39;a&#39;, &#39;a&#39;,</span>
<span class="go"> &#39;c&#39;, &#39;t&#39;, &#39;a&#39;, &#39;c&#39;, &#39;a&#39;, &#39;c&#39;, &#39;t&#39;, &#39;a&#39;, &#39;c&#39;, &#39;t&#39;, &#39;g&#39;, &#39;c&#39;, &#39;c&#39;, &#39;g&#39;, &#39;c&#39;,</span>
<span class="go"> &#39;g&#39;, &#39;c&#39;, &#39;g&#39;, &#39;g&#39;, &#39;c&#39;, &#39;a&#39;, &#39;t&#39;, &#39;a&#39;, &#39;t&#39;, &#39;a&#39;, &#39;t&#39;, &#39;t&#39;, &#39;t&#39;, &#39;a&#39;, &#39;a&#39;,</span>
<span class="go"> &#39;a&#39;, &#39;t&#39;, &#39;a&#39;, &#39;t&#39;, &#39;a&#39;]</span>
</pre></div>
</div>
<p>Python’s list <a class="reference internal" href="glossary.html#term-class"><span class="xref std std-term">class</span></a> has got a method called <code class="xref py py-func docutils literal"><span class="pre">count()</span></code> that we can use
to find out the counts of particular elements in the list.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dna_list</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;a&quot;</span><span class="p">)</span>
<span class="go">17</span>
</pre></div>
</div>
<p>To find out the total number of items in a list one can use Python’s built-in
<code class="xref py py-func docutils literal"><span class="pre">len()</span></code> function, which returns the <em>length</em> of the list.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">dna_list</span><span class="p">)</span>
<span class="go">50</span>
</pre></div>
</div>
<p>When using Python you need to be careful when dividing integers, because in Python 2
the default is to use integer division, i.e. to discard the remainder.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="mi">3</span> <span class="o">/</span> <span class="mi">2</span>
<span class="go">1</span>
</pre></div>
</div>
<p>One can work around this by ensuring that at least one of the numbers is
represented using floating point.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="mi">3</span> <span class="o">/</span> <span class="mf">2.0</span>
<span class="go">1.5</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="first admonition-title">Avertissement</p>
<p class="last">In Python 3, the behaviour of the division operator has been
changed, and dividing two integers will result in normal division.</p>
</div>
<p>One can convert an integer to a floating point number using Python’s built-in
<code class="xref py py-func docutils literal"><span class="pre">float()</span></code> function.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">float</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
<span class="go">2.0</span>
</pre></div>
</div>
<p>We now have all the information required to calculate the GC-content of the DNA
sequence.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gc_count</span> <span class="o">=</span> <span class="n">dna_list</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;g&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">dna_list</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;c&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gc_fraction</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">gc_count</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">dna_list</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="mi">100</span> <span class="o">*</span> <span class="n">gc_fraction</span>
<span class="go">38.0</span>
</pre></div>
</div>
</div>
<div class="section" id="creating-reusable-functions">
<h2>Creating reusable functions</h2>
<p>Suppose that we wanted to calculate the GC-content for several sequences. In
this case it would be very annoying, and error prone, to have to enter the
commands above into the Python shell manually for each sequence. Rather, it
would be advantageous to be able to create a piece of code that could be called
repeatedly to calculate the GC-content. We can achieve this using the concept of
functions. In other words functions are a means for programmers to avoid repeating
themselves, thus adhering to the <a class="reference internal" href="glossary.html#term-dry"><span class="xref std std-term">DRY</span></a> principle.</p>
<p>Let us create a simple function that adds two items together.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">def</span> <span class="nf">add</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">return</span> <span class="n">a</span> <span class="o">+</span> <span class="n">b</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">add</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="go">5</span>
</pre></div>
</div>
<p>In Python functions are defined using the <code class="docutils literal"><span class="pre">def</span></code> keyword. Note that the
<code class="docutils literal"><span class="pre">def</span></code> keyword is followed by the name of the function. The name of the
function is followed by a parenthesized set of arguments, in this case the
function takes two arguments <code class="docutils literal"><span class="pre">a</span></code> and <code class="docutils literal"><span class="pre">b</span></code>. The end of the function
definition is marked using a colon.</p>
<p>The body of the function, in this example the <code class="docutils literal"><span class="pre">return</span></code> statement, needs to be
indented. The standard in Python is to use four white spaces to indent code
blocks. In this case the function body only contains one line of code. However,
a function can include several indented lines of code.</p>
<div class="admonition warning">
<p class="first admonition-title">Avertissement</p>
<p class="last">Whitespace really matters in Python! If your code is not correctly
aligned you will see <code class="docutils literal"><span class="pre">IndentationError</span></code> messages telling you
that everything is not as it should be. You will also run into
<code class="docutils literal"><span class="pre">IndentationError</span></code> messages if you mix white spaces and tabs.</p>
</div>
<p>Now we can create a function for calculating the GC-content of a sequence.
As with variables explicit trumps succinct in terms of naming.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">def</span> <span class="nf">gc_content</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">gc_count</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;g&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;c&quot;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">gc_fraction</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">gc_count</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
<span class="gp">... </span>    <span class="k">return</span> <span class="mi">100</span> <span class="o">*</span> <span class="n">gc_fraction</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gc_content</span><span class="p">(</span><span class="n">dna_list</span><span class="p">)</span>
<span class="go">38.0</span>
</pre></div>
</div>
</div>
<div class="section" id="list-slicing">
<h2>List slicing</h2>
<p>Suppose that we wanted to look at local variability in GC-content. To achieve
this we would like to be able to select segments of our initial list. This is
known as «&nbsp;slicing&nbsp;», as in slicing up a salami.</p>
<p>In Python slicing uses a <code class="docutils literal"><span class="pre">[start:end]</span></code> syntax that is inclusive for the start
index and exclusive for the end index. To illustrate slicing let us first
create a list to work with.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">zero_to_five</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;zero&quot;</span><span class="p">,</span> <span class="s2">&quot;one&quot;</span><span class="p">,</span> <span class="s2">&quot;two&quot;</span><span class="p">,</span> <span class="s2">&quot;three&quot;</span><span class="p">,</span> <span class="s2">&quot;four&quot;</span><span class="p">,</span> <span class="s2">&quot;five&quot;</span><span class="p">]</span>
</pre></div>
</div>
<p>To get the first two elements we therefore use 0 for the start index, as
Python uses a zero-based indexing system, and 2 for the end index as the
element from the end index is excluded.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">zero_to_five</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span>
<span class="go">[&#39;zero&#39;, &#39;one&#39;]</span>
</pre></div>
</div>
<p>Note that the start position for the slicing is 0 by default so we could just
as well have written.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">zero_to_five</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span>
<span class="go">[&#39;zero&#39;, &#39;one&#39;]</span>
</pre></div>
</div>
<p>To get the last three elements.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">zero_to_five</span><span class="p">[</span><span class="mi">3</span><span class="p">:]</span>
<span class="go">[&#39;three&#39;, &#39;four&#39;, &#39;five&#39;]</span>
</pre></div>
</div>
<p>It is worth noting that we can use negative indices, where -1 represents the
last element. So to get all elements except the first and the last, one could
slice the list using the indices 1 and -1.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">zero_to_five</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="go">[&#39;one&#39;, &#39;two&#39;, &#39;three&#39;, &#39;four&#39;]</span>
</pre></div>
</div>
<p>We can use list slicing to calculate the local GC-content measurements of
our DNA.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gc_content</span><span class="p">(</span><span class="n">dna_list</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span>
<span class="go">40.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gc_content</span><span class="p">(</span><span class="n">dna_list</span><span class="p">[</span><span class="mi">10</span><span class="p">:</span><span class="mi">20</span><span class="p">])</span>
<span class="go">30.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gc_content</span><span class="p">(</span><span class="n">dna_list</span><span class="p">[</span><span class="mi">20</span><span class="p">:</span><span class="mi">30</span><span class="p">])</span>
<span class="go">70.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gc_content</span><span class="p">(</span><span class="n">dna_list</span><span class="p">[</span><span class="mi">30</span><span class="p">:</span><span class="mi">40</span><span class="p">])</span>
<span class="go">50.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gc_content</span><span class="p">(</span><span class="n">dna_list</span><span class="p">[</span><span class="mi">40</span><span class="p">:</span><span class="mi">50</span><span class="p">])</span>
<span class="go">0.0</span>
</pre></div>
</div>
</div>
<div class="section" id="loops">
<h2>Loops</h2>
<p>It can get a bit repetitive, tedious, and error prone specifying all the ranges
manually. A better way to do this is to make use of a loop construct. A loop
allows a program to cycle through the same set of operations a number of times.</p>
<p>In lower level languages <em>while</em> loops are common because they operate in a way
that closely mimic how the hardware works. The code below illustrates a typical
setup of a while loop.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cycle</span> <span class="o">=</span> <span class="mi">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">while</span> <span class="n">cycle</span> <span class="o">&lt;</span> <span class="mi">5</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">print</span><span class="p">(</span><span class="n">cycle</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">cycle</span> <span class="o">=</span> <span class="n">cycle</span> <span class="o">+</span> <span class="mi">1</span>
<span class="gp">...</span>
<span class="go">0</span>
<span class="go">1</span>
<span class="go">2</span>
<span class="go">3</span>
<span class="go">4</span>
</pre></div>
</div>
<p>In the code above Python moves through the commands in the while loop executing
them in order, i.e. printing the value of the <code class="docutils literal"><span class="pre">cycle</span></code> variable and then
incrementing it. The logic then moves back to the <code class="docutils literal"><span class="pre">while</span></code> statement and
the conditional (<code class="docutils literal"><span class="pre">cycle</span> <span class="pre">&lt;</span> <span class="pre">5</span></code>) is re-evaluated. If true the commands in the
while statment are executed in order again, and so forth until the conditional
is false. In this example the <code class="docutils literal"><span class="pre">print(cycle)</span></code> command was called five times,
i.e. until the <code class="docutils literal"><span class="pre">cycle</span></code> variable incremented to 5 and the <code class="docutils literal"><span class="pre">cycle</span> <span class="pre">&lt;</span> <span class="pre">5</span></code>
conditional evaluated to false.</p>
<p>However, when working in Python it is much more common to make use of <em>for</em>
loops. For loops are used to iterate over elements in data structures such as
lists.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="k">print</span><span class="p">(</span><span class="n">item</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">0</span>
<span class="go">1</span>
<span class="go">2</span>
<span class="go">3</span>
<span class="go">4</span>
</pre></div>
</div>
<p>In the above we had to manually write out all the numbers that we wanted. However,
because iterating over a range of integers is such a common task Python has a
built-in function for generating such lists.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
<span class="go">[0, 1, 2, 3, 4]</span>
</pre></div>
</div>
<p>So a typical for loop might look like the below.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">item</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">0</span>
<span class="go">1</span>
<span class="go">2</span>
<span class="go">3</span>
<span class="go">4</span>
</pre></div>
</div>
<p>The <code class="xref py py-func docutils literal"><span class="pre">range()</span></code> function can also be told to start at a larger number. Say for
example that we wanted a list including the numbers 5, 6 and 7.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">8</span><span class="p">)</span>
<span class="go">[5, 6, 7]</span>
</pre></div>
</div>
<p>As with slicing the start value is included whereas the end value is excluded.</p>
<p>It is also possible to alter the step size. To do this we must specify the start
and end values explicitly before adding the step size.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span>
<span class="go">[0, 10, 20, 30, 40]</span>
</pre></div>
</div>
<p>We are now in a position where we can create a naive loop for for calculating
the local GC-content of our DNA.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">start</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">10</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">end</span> <span class="o">=</span> <span class="n">start</span> <span class="o">+</span> <span class="mi">10</span>
<span class="gp">... </span>    <span class="k">print</span><span class="p">(</span><span class="n">gc_content</span><span class="p">(</span><span class="n">dna_list</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]))</span>
<span class="gp">...</span>
<span class="go">40.0</span>
<span class="go">30.0</span>
<span class="go">70.0</span>
<span class="go">50.0</span>
<span class="go">0.0</span>
</pre></div>
</div>
<p>Loops are really powerful. They provide a means to iterate over lots of items
and can be used to automate repetitive tasks.</p>
</div>
<div class="section" id="creating-a-sliding-window-gc-content-function">
<h2>Creating a sliding window GC-content function</h2>
<p>So far we have been working with Python in interactive mode. This is a great
way to explore what can be achieved with Python.  It is also handy to simply
use Python’s interactive mode as a command line calculator.  However, it can
get a little bit clunky when trying to write constructs that span several
lines, such as functions.</p>
<p>Now we will examine how one can write a Python script as a text file and how to
run that text file through the Python interpreter, i.e. how to run a Python
script from the command line.</p>
<p>Start off by creating a new directory to work in.</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>$ mkdir S.coelicolor-local-GC-content
$ cd S.coelicolor-local-GC-content
</pre></div>
</div>
<p>Use your favourite text editor to enter the code below into a file
named <code class="docutils literal"><span class="pre">gc_content.py</span></code>.</p>
<div class="highlight-python"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><span class="n">sequence</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="s2">&quot;attagcgcaatctaactacactactgccgcgcggcatatatttaaatata&quot;</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
</pre></div>
</td></tr></table></div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If your text editor is not giving you syntax highlighting find out
how it can be enabled. If your text editor does not support syntax
highlighting find a better text editor!</p>
</div>
<p>Open up a terminal and go to the directory where you saved the <code class="docutils literal"><span class="pre">gc_content.py</span></code>
script. Run the script using the command below.</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>$ python gc_content.py
</pre></div>
</div>
<p>You should see the output below printed to your terminal.</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>[&#39;a&#39;, &#39;t&#39;, &#39;t&#39;, &#39;a&#39;, &#39;g&#39;, &#39;c&#39;, &#39;g&#39;, &#39;c&#39;, &#39;a&#39;, &#39;a&#39;, &#39;t&#39;, &#39;c&#39;, &#39;t&#39;, &#39;a&#39;, &#39;a&#39;,
&#39;c&#39;, &#39;t&#39;, &#39;a&#39;, &#39;c&#39;, &#39;a&#39;, &#39;c&#39;, &#39;t&#39;, &#39;a&#39;, &#39;c&#39;, &#39;t&#39;, &#39;g&#39;, &#39;c&#39;, &#39;c&#39;, &#39;g&#39;, &#39;c&#39;,
&#39;g&#39;, &#39;c&#39;, &#39;g&#39;, &#39;g&#39;, &#39;c&#39;, &#39;a&#39;, &#39;t&#39;, &#39;a&#39;, &#39;t&#39;, &#39;a&#39;, &#39;t&#39;, &#39;t&#39;, &#39;t&#39;, &#39;a&#39;, &#39;a&#39;,
&#39;a&#39;, &#39;t&#39;, &#39;a&#39;, &#39;t&#39;, &#39;a&#39;]
</pre></div>
</div>
<p>In the script we used Python’s built-in <code class="xref py py-func docutils literal"><span class="pre">list()</span></code> function to convert the
DNA string into a list. We then printed out the <code class="docutils literal"><span class="pre">sequence</span></code> list.</p>
<p>Now let us add the <code class="xref py py-func docutils literal"><span class="pre">gc_content()</span></code> function to the script.</p>
<div class="highlight-python"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5
6
7
8</pre></div></td><td class="code"><div class="highlight"><pre><span></span><span class="hll"><span class="k">def</span> <span class="nf">gc_content</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
</span><span class="hll">    <span class="s2">&quot;Return GC-content as a percentage from a list of DNA letters.&quot;</span>
</span><span class="hll">    <span class="n">gc_count</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;g&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;c&quot;</span><span class="p">)</span>
</span><span class="hll">    <span class="n">gc_fraction</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">gc_count</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
</span><span class="hll">    <span class="k">return</span> <span class="mi">100</span> <span class="o">*</span> <span class="n">gc_fraction</span>
</span>
<span class="n">sequence</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="s2">&quot;attagcgcaatctaactacactactgccgcgcggcatatatttaaatata&quot;</span><span class="p">)</span>
<span class="hll"><span class="k">print</span><span class="p">(</span><span class="n">gc_content</span><span class="p">(</span><span class="n">sequence</span><span class="p">))</span>
</span></pre></div>
</td></tr></table></div>
<p>In the above the <code class="xref py py-func docutils literal"><span class="pre">gc_content()</span></code> function is implemented as per our
exploration in our interactive session. The only difference is the
addition of a, so called, «&nbsp;docstring&nbsp;» (documentation string) to the body
of the function (line 2). The docstring is meant to document the purpose
and usage of the function. Documenting the code in this way makes it
easier for you, and others, to understand it.</p>
<p>Note that the script now prints out the GC-content rather than the sequence
(line 8). Let us run the updated script from the command line.</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>$ python gc_content.py
38.0
</pre></div>
</div>
<p>The next piece of code will be a bit more complicated. However, note that
that it represents the most complicated aspect of this chapter. So if you
find it difficult, don’t give up, it gets easier again later on.</p>
<p>Now let us implement a new function for performing a sliding window analysis.
Add the code below to the start of the <code class="docutils literal"><span class="pre">gc_content.py</span></code> file.</p>
<div class="highlight-python"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12</pre></div></td><td class="code"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">sliding_window_analysis</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">function</span><span class="p">,</span> <span class="n">window_size</span><span class="o">=</span><span class="mi">10</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return an iterator that yields (start, end, property) tuples.</span>

<span class="sd">    Where start and end are the indices used to slice the input list</span>
<span class="sd">    and property is the return value of the function given the sliced</span>
<span class="sd">    list.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">start</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">),</span> <span class="n">window_size</span><span class="p">):</span>
        <span class="n">end</span> <span class="o">=</span> <span class="n">start</span> <span class="o">+</span> <span class="n">window_size</span>
        <span class="k">if</span> <span class="n">end</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
            <span class="k">break</span>
        <span class="k">yield</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">sequence</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">])</span>
</pre></div>
</td></tr></table></div>
<p>There is quite a lot going on in the code above so let us walk through it
slowly. One of the first things to note is that the
<code class="xref py py-func docutils literal"><span class="pre">sliding_window_analysis()</span></code> function takes another <code class="docutils literal"><span class="pre">function</span></code> as its
second argument. Functions can be passed around just like variables and on line
12 the <code class="docutils literal"><span class="pre">function</span></code> is repeatedly called with slices of the input sequence.</p>
<p>The <code class="xref py py-func docutils literal"><span class="pre">sliding_window_analysis()</span></code> function also takes a <code class="docutils literal"><span class="pre">window_size</span></code>
argument. This defines the step size of the <code class="xref py py-func docutils literal"><span class="pre">range()</span></code> function used to
generate the <code class="docutils literal"><span class="pre">start</span></code> indices for the slicing. Note that in this case we
provide the <code class="docutils literal"><span class="pre">window_size</span></code> argument with a default value of 10. This means
that the <code class="docutils literal"><span class="pre">window_size</span></code> argument does not need to be explicitly set when
calling the function (if one is happy with the default).</p>
<p>On line 9, inside the for loop, we generate the <code class="docutils literal"><span class="pre">end</span></code> index by adding the
<code class="docutils literal"><span class="pre">window_size</span></code> to the <code class="docutils literal"><span class="pre">start</span></code> index. This is followed by a check that the
generated <code class="docutils literal"><span class="pre">end</span></code> index would not result in a list slice that spanned beyond
the end of the sequence.</p>
<p>On line 11, inside the <code class="docutils literal"><span class="pre">if</span></code> statment there is a <code class="docutils literal"><span class="pre">break</span></code> clause. The <code class="docutils literal"><span class="pre">break</span></code>
statement is used to break out of the loop immediately. In other words if the
<code class="docutils literal"><span class="pre">end</span></code> variable is greater than the length of the sequence we break out of the
loop immediately.</p>
<p>At the end of the for loop we make use of the <code class="docutils literal"><span class="pre">yield</span></code> keyword to pass on the
<code class="docutils literal"><span class="pre">start</span></code> and <code class="docutils literal"><span class="pre">end</span></code> indices as well as the value resulting from calling the
input <code class="docutils literal"><span class="pre">function</span></code> with the sequence slice.  This means that rather than
returning a value the <code class="xref py py-func docutils literal"><span class="pre">sliding_window_analysis()</span></code> function returns an
iterator. As the name suggests an «&nbsp;iterator&nbsp;» is an object that one can iterate
over, for example using a <em>for</em> loop. Let us add some code to the script to
illustrate how one would use the <code class="xref py py-func docutils literal"><span class="pre">sliding_window_analysis()</span></code> function in
practise.</p>
<div class="highlight-python"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre>20
21
22</pre></div></td><td class="code"><div class="highlight"><pre><span></span><span class="n">sequence</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="s2">&quot;attagcgcaatctaactacactactgccgcgcggcatatatttaaatata&quot;</span><span class="p">)</span>
<span class="hll"><span class="k">for</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">gc</span> <span class="ow">in</span> <span class="n">sliding_window_analysis</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">gc_content</span><span class="p">):</span>
</span><span class="hll">    <span class="k">print</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">gc</span><span class="p">)</span>
</span></pre></div>
</td></tr></table></div>
<p>Let us test the code again.</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>$ python gc_content.py
(0, 10, 40.0)
(10, 20, 30.0)
(20, 30, 70.0)
(30, 40, 50.0)
(40, 50, 0.0)
</pre></div>
</div>
<p>The current implementation of the <code class="xref py py-func docutils literal"><span class="pre">sliding_window_analysis()</span></code> is very
dependent on the frame of reference as the window slides along. For example if
the <code class="docutils literal"><span class="pre">window_size</span></code> argument was set to 3 one would obtain the analysis of the
first codon reading frame, but one would have no information about the second
and third codon reading frames. To overcome this one can perform sliding window
analysis with overlapping windows. Let us illustrate this visually by extracting
codons from a DNA sequence.</p>
<div class="highlight-none"><div class="highlight"><pre><span></span># Original sequence.
atcgctaaa

# Non overlapping windows.
atc
   gct
      aaa

# Overlapping windows.
atc
 tcg
  cgc
   gct
    cta
     taa
      aaa
</pre></div>
</div>
<p>To enable overlapping windows in our <code class="xref py py-func docutils literal"><span class="pre">sliding_window_analysis()</span></code> function
we need to add a <code class="docutils literal"><span class="pre">step_size</span></code> argument to it and make use of this in the call
to the <code class="xref py py-func docutils literal"><span class="pre">range()</span></code> function.</p>
<div class="highlight-python"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12</pre></div></td><td class="code"><div class="highlight"><pre><span></span><span class="hll"><span class="k">def</span> <span class="nf">sliding_window_analysis</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">function</span><span class="p">,</span> <span class="n">window_size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">step_size</span><span class="o">=</span><span class="mi">5</span><span class="p">):</span>
</span>    <span class="sd">&quot;&quot;&quot;Return an iterator that yields (start, end, property) tuples.</span>

<span class="sd">    Where start and end are the indices used to slice the input list</span>
<span class="sd">    and property is the return value of the function given the sliced</span>
<span class="sd">    list.</span>
<span class="sd">    &quot;&quot;&quot;</span>
<span class="hll">    <span class="k">for</span> <span class="n">start</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">),</span> <span class="n">step_size</span><span class="p">):</span>
</span>        <span class="n">end</span> <span class="o">=</span> <span class="n">start</span> <span class="o">+</span> <span class="n">window_size</span>
        <span class="k">if</span> <span class="n">end</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
            <span class="k">break</span>
        <span class="k">yield</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">sequence</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">])</span>
</pre></div>
</td></tr></table></div>
<p>Let us run the script again to see what the output of this overlapping sliding
window analysis is.</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>$ python gc_content.py
(0, 10, 40.0)
(5, 15, 40.0)
(10, 20, 30.0)
(15, 25, 40.0)
(20, 30, 70.0)
(25, 35, 100.0)
(30, 40, 50.0)
(35, 45, 0.0)
(40, 50, 0.0)
</pre></div>
</div>
<p>Note that the <code class="xref py py-func docutils literal"><span class="pre">gc_content()</span></code> function is now applied to overlapping
segments of DNA. This allows us, for example, to note that the 25 to 35 region
has a GC-content of 100%, which is something that we did not manage to pick out
before.</p>
</div>
<div class="section" id="downloading-the-genome">
<h2>Downloading the genome</h2>
<p>It is time to start working on some real data. Let us download the genome of
<em>Streptomyces coelicolor</em> from the <a class="reference external" href="ftp://ftp.sanger.ac.uk/pub/project/pathogens/S_coelicolor/whole_genome/">Sanger Centre ftp site</a>. The
URL shortened using <a class="reference external" href="https://bitly.com/">bitly</a> point to the <code class="docutils literal"><span class="pre">Sco.dna</span></code> file.</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>$ curl --location --output Sco.dna http://bit.ly/1Q8eKWT
$ head Sco.dna
SQ   Sequence 8667507 BP; 1203558 A; 3121252 C; 3129638 G; 1213059 T; 0 other;
     cccgcggagc gggtaccaca tcgctgcgcg atgtgcgagc gaacacccgg gctgcgcccg        60
     ggtgttgcgc tcccgctccg cgggagcgct ggcgggacgc tgcgcgtccc gctcaccaag       120
     cccgcttcgc gggcttggtg acgctccgtc cgctgcgctt ccggagttgc ggggcttcgc       180
     cccgctaacc ctgggcctcg cttcgctccg ccttgggcct gcggcgggtc cgctgcgctc       240
     ccccgcctca agggcccttc cggctgcgcc tccaggaccc aaccgcttgc gcgggcctgg       300
     ctggctacga ggatcggggg tcgctcgttc gtgtcgggtt ctagtgtagt ggctgcctca       360
     gatagatgca gcatgtatcg ttggcagaaa tatgggacac ccgccagtca ctcgggaatc       420
     tcccaagttt cgagaggatg gccagatgac cggtcaccac gaatctaccg gaccaggtac       480
     cgcgctgagc agcgattcga cgtgccgggt gacgcagtat cagacggcgg gtgtgaacgc       540
</pre></div>
</div>
</div>
<div class="section" id="reading-and-writing-files">
<h2>Reading and writing files</h2>
<p>In order to be able to process the genome of <em>Streptomyces coelicolor</em> we need
to be able to read in the <code class="docutils literal"><span class="pre">Sco.dna</span></code> file. In Python reading and writing of files
is achieved using the built-in <code class="xref py py-func docutils literal"><span class="pre">open()</span></code> function, which returns a file handle.</p>
<div class="sidebar">
<p class="first sidebar-title">What is a file handle?</p>
<p class="last">A file handle is a data structure that handles the book keeping of the
position within the file as well as the mode in which the file was opened.
The mode of the file determines what one can do with it. For example one
cannot write to a file that has been opened for reading. The position of
the file handle determines where the next operation will take place. For
example, if one is about to write something to a file existing content will
be overwritten unless the position is pointing at the end of the file, in
which case the new content will be appended to the old.</p>
</div>
<p>Before we start adding code to our script let us examine reading and writing of files
using in Python’s interactive mode. Let us open up the <code class="docutils literal"><span class="pre">Sco.dna</span></code> file for reading.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">file_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;Sco.dna&quot;</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;r&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>We can access the current position within the file using the <code class="xref py py-func docutils literal"><span class="pre">tell()</span></code> method of
the file handle.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">file_handle</span><span class="o">.</span><span class="n">tell</span><span class="p">()</span>
<span class="go">0</span>
</pre></div>
</div>
<p>The integer zero indicates that we are at the beginning of the file.</p>
<p>To read in the entire content of the file as a single string of text one can use the
<code class="xref py py-func docutils literal"><span class="pre">read()</span></code> method of the file handle.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">text</span> <span class="o">=</span> <span class="n">file_handle</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
</pre></div>
</div>
<p>After having read in the content of the file the position of the file handle
will point at the end of the file.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">file_handle</span><span class="o">.</span><span class="n">tell</span><span class="p">()</span>
<span class="go">11701261</span>
</pre></div>
</div>
<p>When one has finished working with a file handle it is important to remember to
close it.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">file_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>Let us examine the text that we read in.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">type</span><span class="p">(</span><span class="n">text</span><span class="p">)</span>
<span class="go">&lt;type &#39;str&#39;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">text</span><span class="p">)</span>
<span class="go">11701261</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">text</span><span class="p">[:</span><span class="mi">60</span><span class="p">]</span>
<span class="go">&#39;SQ   Sequence 8667507 BP; 1203558 A; 3121252 C; 3129638 G; 1&#39;</span>
</pre></div>
</div>
<p>However, rather than reading in files as continuous strings one often want to process
files line by line. One can read in a file as a list of lines using the <code class="xref py py-func docutils literal"><span class="pre">readlines()</span></code>
method.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">file_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;Sco.dna&quot;</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lines</span> <span class="o">=</span> <span class="n">file_handle</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>Let us examine the lines that we read in.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">type</span><span class="p">(</span><span class="n">lines</span><span class="p">)</span>
<span class="go">&lt;type &#39;list&#39;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">lines</span><span class="p">)</span>
<span class="go">144461</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lines</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">&#39;SQ   Sequence 8667507 BP; 1203558 A; 3121252 C; 3129638 G; 1213059 T; 0 other;\n&#39;</span>
</pre></div>
</div>
<p>A third way of accessing the content of a file handle is to simply treat it as
an iterator. This is possible as the Python file handles implement a method called
<code class="xref py py-func docutils literal"><span class="pre">next()</span></code> that returns the next line in the file. When it reaches the end of the
file the <code class="xref py py-func docutils literal"><span class="pre">next()</span></code> function raises a <code class="docutils literal"><span class="pre">StopIteration</span></code> exception, which tells the
iterator to stop iterating.</p>
<p>Let’s see the workings of the <code class="xref py py-func docutils literal"><span class="pre">next()</span></code> method in action.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">file_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;Sco.dna&quot;</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_handle</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>
<span class="go">&#39;SQ   Sequence 8667507 BP; 1203558 A; 3121252 C; 3129638 G; 1213059 T; 0 other;\n&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_handle</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>
<span class="go">&#39;     cccgcggagc gggtaccaca tcgctgcgcg atgtgcgagc gaacacccgg gctgcgcccg        60\n&#39;</span>
</pre></div>
</div>
<p>We can go to the end of the file using the <code class="xref py py-func docutils literal"><span class="pre">seek()</span></code> method of the file handle.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">file_handle</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">11701261</span><span class="p">)</span>
</pre></div>
</div>
<p>Let’s see what happens when we call the <code class="xref py py-func docutils literal"><span class="pre">next()</span></code> method now.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">file_handle</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>
<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;&lt;stdin&gt;&quot;</span>, line <span class="m">1</span>, in <span class="n">&lt;module&gt;</span>
<span class="gr">StopIteration</span>
</pre></div>
</div>
<p>As explained above this raises a <code class="docutils literal"><span class="pre">StopIteration</span></code> exception. Now that we are done
with our experiment we must remember to close the file handle.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">file_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>In practise one tends to use file handles directly within for loops.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">num_lines</span> <span class="o">=</span> <span class="mi">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;Sco.dna&quot;</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">file_handle</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">num_lines</span> <span class="o">=</span> <span class="n">num_lines</span> <span class="o">+</span> <span class="mi">1</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">num_lines</span><span class="p">)</span>
<span class="go">144461</span>
</pre></div>
</div>
<p>In the for loop above the file handle acts as an iterator, yielding
the lines within the opened file.</p>
<p>Again we must not forget to close the file handle.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">file_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>Having to constantly remember to close file handles when one is done with them
can become tedious. Furthermore, forgetting to close file handles can have dire
consequences. To make life easier one can make use of Python’s built-in <code class="docutils literal"><span class="pre">with</span></code>
keyword.</p>
<p>The <code class="docutils literal"><span class="pre">with</span></code> keywords works with context managers. A context
manager implements the so called «&nbsp;context manager protocol&nbsp;». In the case of a
file handle this means that the file is opened when one enters into the context
of the <code class="docutils literal"><span class="pre">with</span></code> statement and that the file is automatically closed when one
exits out of the context. All that is a fancy way of saying that we do not need
to worry about remembering to close files if we access file handles using the
syntax below.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;Sco.dna&quot;</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file_handle</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">text</span> <span class="o">=</span> <span class="n">file_handle</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="gp">...</span>
</pre></div>
</div>
<p>Let us shift the focus to the writing of files. There are two modes for writing
files <code class="docutils literal"><span class="pre">w</span></code> and <code class="docutils literal"><span class="pre">a</span></code>. The former will overwrite any existing files with the
same name whereas the latter would append to them. Let us illustrate this with
an example.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;tmp.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file_handle</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">file_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;Original message&quot;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;tmp.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file_handle</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">print</span><span class="p">(</span><span class="n">file_handle</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
<span class="gp">...</span>
<span class="go">Original message</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;tmp.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;a&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file_handle</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">file_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;, with more info&quot;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;tmp.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file_handle</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">print</span><span class="p">(</span><span class="n">file_handle</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
<span class="gp">...</span>
<span class="go">Original message, with more info</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;tmp.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file_handle</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">file_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;scrap that...&quot;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;tmp.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file_handle</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">print</span><span class="p">(</span><span class="n">file_handle</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
<span class="gp">...</span>
<span class="go">scrap that...</span>
</pre></div>
</div>
<p>Armed with our new found knowledge of how to read and write files we will now
create a function for reading in the DNA sequence from the <code class="docutils literal"><span class="pre">Sco.dna</span></code> file.</p>
</div>
<div class="section" id="creating-a-function-for-reading-in-the-streptomyces-sequence">
<h2>Creating a function for reading in the <em>Streptomyces</em> sequence</h2>
<p>Let us create a function that reads in a genome as a string and
returns the DNA sequence as a list. At this
point we have a choice of what the input parameter should be. We could give the
function the name of the file containing the genome or we could give the
function a file handle of the genome. Personally, I prefer to create functions
that accept file handles, because they are more generic. Sometimes the data to
be read comes from sources other than a file on disk. However, as long as these
behave as a file object one can still pass them to the function.</p>
<p>Let us have a look at the file containing the <em>Streptomyces coelicolor</em> genome.</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>$ head Sco.dna
SQ   Sequence 8667507 BP; 1203558 A; 3121252 C; 3129638 G; 1213059 T; 0 other;
     cccgcggagc gggtaccaca tcgctgcgcg atgtgcgagc gaacacccgg gctgcgcccg        60
     ggtgttgcgc tcccgctccg cgggagcgct ggcgggacgc tgcgcgtccc gctcaccaag       120
     cccgcttcgc gggcttggtg acgctccgtc cgctgcgctt ccggagttgc ggggcttcgc       180
     cccgctaacc ctgggcctcg cttcgctccg ccttgggcct gcggcgggtc cgctgcgctc       240
     ccccgcctca agggcccttc cggctgcgcc tccaggaccc aaccgcttgc gcgggcctgg       300
     ctggctacga ggatcggggg tcgctcgttc gtgtcgggtt ctagtgtagt ggctgcctca       360
     gatagatgca gcatgtatcg ttggcagaaa tatgggacac ccgccagtca ctcgggaatc       420
     tcccaagttt cgagaggatg gccagatgac cggtcaccac gaatctaccg gaccaggtac       480
     cgcgctgagc agcgattcga cgtgccgggt gacgcagtat cagacggcgg gtgtgaacgc       540
</pre></div>
</div>
<p>From this we want a function that:</p>
<ol class="arabic simple">
<li>Discards the first line, as it does not contain any sequence</li>
<li>Iterates over all subsequent lines extracting the relevant sequence from them</li>
</ol>
<p>Extracting the relevant sequence can be achieved by noting that each sequence
line consists of seven «&nbsp;words&nbsp;», where a word is defined as a set of characters
separated by one or more white spaces. The first six words correspond to sequence,
whereas the last word is an index listing the number of nucleotide bases.</p>
<p>Let us implement such a function. Add the lines below to the top of the
<code class="docutils literal"><span class="pre">gc_content.py</span></code> file.</p>
<div class="highlight-python"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10</pre></div></td><td class="code"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">parse_dna</span><span class="p">(</span><span class="n">file_handle</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return DNA sequence as a list.&quot;&quot;&quot;</span>
    <span class="n">first_line</span> <span class="o">=</span> <span class="n">file_handle</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>  <span class="c1"># Discard the first line.</span>
    <span class="n">sequence</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">file_handle</span><span class="p">:</span>
        <span class="n">words</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
        <span class="n">seq_string</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">words</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">seq_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">seq_string</span><span class="p">)</span>
        <span class="n">sequence</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">seq_list</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">sequence</span>
</pre></div>
</td></tr></table></div>
<p>There are a couple of new string methods introduced in the above, let’s explain
them now.</p>
<p>Let’s look at line six first.</p>
<div class="highlight-python"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre>6</pre></div></td><td class="code"><div class="highlight"><pre><span></span>        <span class="n">words</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
</pre></div>
</td></tr></table></div>
<p>Here we use the <code class="xref py py-func docutils literal"><span class="pre">split()</span></code> method to split the string into
a list of words, by default the <code class="xref py py-func docutils literal"><span class="pre">split()</span></code> method splits text based on one
or more white space characters.</p>
<p>On line seven we use the <code class="xref py py-func docutils literal"><span class="pre">join()</span></code> method to join the words together.</p>
<div class="highlight-python"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre>7</pre></div></td><td class="code"><div class="highlight"><pre><span></span>        <span class="n">seq_string</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">words</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
</pre></div>
</td></tr></table></div>
<p>In this instance there are no characters separating the words to be joined. It
is worth clarifying this with an example, if we wanted to join the words using
a comma character one would use the syntax <code class="docutils literal"><span class="pre">&quot;,&quot;.join(words[:-1])</span></code>.</p>
<p>On line seven it is also worth noting that we exclude the last word (the
numerical index) by making use of list slicing <code class="docutils literal"><span class="pre">words[:-1]</span></code>.</p>
<p>Finally, on line nine we make use of the list method <code class="xref py py-func docutils literal"><span class="pre">extend()</span></code>, this
extends the existing <code class="docutils literal"><span class="pre">sequence</span></code> list with all the elements from the
<code class="docutils literal"><span class="pre">seq_list</span></code> list. Because <code class="docutils literal"><span class="pre">words</span></code> <code class="docutils literal"><span class="pre">seq_string</span></code> and <code class="docutils literal"><span class="pre">seq_list</span></code> will be
overwritten when the loop moves on to the next line in the input file.</p>
<div class="highlight-python"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre>9</pre></div></td><td class="code"><div class="highlight"><pre><span></span>        <span class="n">sequence</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">seq_list</span><span class="p">)</span>
</pre></div>
</td></tr></table></div>
<div class="sidebar">
<p class="first sidebar-title">Adding elements to lists <code class="docutils literal"><span class="pre">append()</span></code> vs <code class="docutils literal"><span class="pre">extend()</span></code></p>
<p>There are two main ways of adding elements to a list. The first is
the append method, which adds a single element.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="s2">&quot;green&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">colors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;blue&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The <code class="docutils literal"><span class="pre">extend()</span></code> method is different in that it extends the list
with the content of another list.</p>
<div class="last highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">colors</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="s2">&quot;yellow&quot;</span><span class="p">,</span> <span class="s2">&quot;purple&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">colors</span><span class="p">)</span>
<span class="go">[&#39;red&#39;, &#39;green&#39;, &#39;blue&#39;, &#39;yellow&#39;, &#39;purple&#39;]</span>
</pre></div>
</div>
</div>
<p>Now let us update the <code class="docutils literal"><span class="pre">gc_content.py</span></code> script to initalise the sequence by
parsing the <code class="docutils literal"><span class="pre">Sco.dna</span></code> file.</p>
<div class="highlight-python"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre>31
32
33
34
35</pre></div></td><td class="code"><div class="highlight"><pre><span></span><span class="hll"><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;Sco.dna&quot;</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file_handle</span><span class="p">:</span>
</span><span class="hll">    <span class="n">sequence</span> <span class="o">=</span> <span class="n">parse_dna</span><span class="p">(</span><span class="n">file_handle</span><span class="p">)</span>
</span>
<span class="k">for</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">gc</span> <span class="ow">in</span> <span class="n">sliding_window_analysis</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">gc_content</span><span class="p">):</span>
    <span class="k">print</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">gc</span><span class="p">)</span>
</pre></div>
</td></tr></table></div>
<p>Finally, let us change the default <code class="docutils literal"><span class="pre">window_size</span></code> and <code class="docutils literal"><span class="pre">step_size</span></code> values. In
the below I have split the function definition over two lines so as not to make
the line exceed 78 characters. Exceeding 78 characters is considered poor «&nbsp;style&nbsp;»
because it makes it difficult to read the code.</p>
<div class="highlight-python"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre>12
13
14
15
16
17
18
19</pre></div></td><td class="code"><div class="highlight"><pre><span></span><span class="hll"><span class="k">def</span> <span class="nf">sliding_window_analysis</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">function</span><span class="p">,</span>
</span><span class="hll">                            <span class="n">window_size</span><span class="o">=</span><span class="mi">100000</span><span class="p">,</span> <span class="n">step_size</span><span class="o">=</span><span class="mi">50000</span><span class="p">):</span>
</span>    <span class="sd">&quot;&quot;&quot;Return an iterator that yields (start, end, property) tuples.</span>

<span class="sd">    Where start and end are the indices used to slice the input list</span>
<span class="sd">    and property is the return value of the function given the sliced</span>
<span class="sd">    list.</span>
<span class="sd">    &quot;&quot;&quot;</span>
</pre></div>
</td></tr></table></div>
<p>Let us run the script again.</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>$ python gc_content.py
</pre></div>
</div>
<p>Note that this will produce a lot of output. To find out the number of lines
that are generated we can make use of piping and the <code class="docutils literal"><span class="pre">wc</span> <span class="pre">-l</span></code> command
(mnemonic <code class="docutils literal"><span class="pre">wc</span></code> word count, <code class="docutils literal"><span class="pre">-l</span></code> lines) .</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>$ python gc_content.py | wc -l
    173
</pre></div>
</div>
<p>The result makes sense as there are 8667507 base pairs in the sequence and
we are stepping through it using a step size of 50000.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="mi">8667507</span> <span class="o">/</span> <span class="mi">50000</span>
<span class="go">173</span>
</pre></div>
</div>
</div>
<div class="section" id="writing-out-the-sliding-window-analysis">
<h2>Writing out the sliding window analysis</h2>
<p>Finally we will write out the analysis to a text file. Since this data is tabular
we will use the CSV file format. Furthermore, since we will want to plot the data using ggplot2 in <a class="reference internal" href="data-visualisation.html"><span class="doc">Data visualisation</span></a> we will use a «&nbsp;tidy&nbsp;» data structure, see <a class="reference internal" href="structuring-and-storing-data.html#tidy-data"><span class="std std-ref">Tidy data</span></a> for details.</p>
<p>Edit the end of the <code class="docutils literal"><span class="pre">gc_content.py</span></code> script to make it look like the below.</p>
<div class="highlight-python"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre>32
33
34
35
36
37
38
39
40
41</pre></div></td><td class="code"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;Sco.dna&quot;</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file_handle</span><span class="p">:</span>
    <span class="n">sequence</span> <span class="o">=</span> <span class="n">parse_dna</span><span class="p">(</span><span class="n">file_handle</span><span class="p">)</span>

<span class="hll"><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;local_gc_content.csv&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file_handle</span><span class="p">:</span>
</span><span class="hll">    <span class="n">header</span> <span class="o">=</span> <span class="s2">&quot;start,middle,end,gc_content</span><span class="se">\n</span><span class="s2">&quot;</span>
</span><span class="hll">    <span class="n">file_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">header</span><span class="p">)</span>
</span><span class="hll">    <span class="k">for</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">gc</span> <span class="ow">in</span> <span class="n">sliding_window_analysis</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">gc_content</span><span class="p">):</span>
</span><span class="hll">        <span class="n">middle</span> <span class="o">=</span> <span class="p">(</span><span class="n">start</span> <span class="o">+</span> <span class="n">end</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span>
</span><span class="hll">        <span class="n">row</span> <span class="o">=</span> <span class="s2">&quot;{},{},{},{}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">middle</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">gc</span><span class="p">)</span>
</span><span class="hll">        <span class="n">file_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
</span></pre></div>
</td></tr></table></div>
<p>On line 35 we open a file handle to write to. On lines 36 and 37 we write a header
to the CSV file. Lines 38 to 41 then performs the sliding window analysis and
writes the results as rows, or lines if you prefer, to the CSV file.
Line 39 calculates the middle of the local sequence by calculating the mean of the
start and end positions.</p>
<p>The main new feature introduced in the code snippet above is on line 40 where
we use Python’s built in string formatting functionality. The matching curly
braces in the string are replaced with the content of the <code class="xref py py-meth docutils literal"><span class="pre">format()</span></code> string
method. Let us illustrate this with an example in interactive mode.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="s2">&quot;{},{},{},{}&quot;</span><span class="p">)</span>
<span class="go">{},{},{},{}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="s2">&quot;{},{},{},{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mf">38.5</span><span class="p">))</span>
<span class="go">1,5,10,38.5</span>
</pre></div>
</div>
<p>Okay, let us see what happens when we run the script.</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>$ python gc_content.py
</pre></div>
</div>
<p>This should have created a file named <code class="docutils literal"><span class="pre">local_gc_content.csv</span></code> in the working directory.</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>$ ls
Sco.dna
gc_content.py
local_gc_content.csv
</pre></div>
</div>
<p>We can examine the top of this newly created file using the <code class="docutils literal"><span class="pre">head</span></code> command.</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>$ head local_gc_content.csv
start,middle,end,gc_content
0,50000,100000,69.124
50000,100000,150000,70.419
100000,150000,200000,72.495
150000,200000,250000,71.707
200000,250000,300000,71.098
250000,300000,350000,72.102
300000,350000,400000,72.712
350000,400000,450000,73.15
400000,450000,500000,73.27
</pre></div>
</div>
<p>Great, everything seems to be working. Let’s start tracking our code using Git,
see <a class="reference internal" href="keeping-track-of-your-work.html"><span class="doc">Keeping track of your work</span></a>.</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>$ git init
$ git status
On branch master

Initial commit

Untracked files:
  (use &quot;git add &lt;file&gt;...&quot; to include in what will be committed)

        Sco.dna
        gc_content.py
        local_gc_content.csv

nothing added to commit but untracked files present (use &quot;git add&quot; to track)
</pre></div>
</div>
<p>We have got three untracked files in our directory, the script, the input data and the output data. We don’t want to track the input and the output data so let’s create a <code class="docutils literal"><span class="pre">.gitignore</span></code> file and add those files to it.</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>Sco.dna
local_gc_content.csv
</pre></div>
</div>
<p>Let’s check the status of our Git repository again.</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>$ git status
On branch master

Initial commit

Untracked files:
  (use &quot;git add &lt;file&gt;...&quot; to include in what will be committed)

        .gitignore
        gc_content.py

nothing added to commit but untracked files present (use &quot;git add&quot; to track)
</pre></div>
</div>
<p>Let’s start tracking the <code class="docutils literal"><span class="pre">gc_content.py</span></code> and the <code class="docutils literal"><span class="pre">.gitignore</span></code> files and take a snapshot of them in their current form.</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>$ git add gc_content.py .gitignore
$ git commit -m &quot;Initial file import&quot;
[master (root-commit) 6d8e0cf] Initial file import
 2 files changed, 43 insertions(+)
 create mode 100644 .gitignore
 create mode 100644 gc_content.py
</pre></div>
</div>
<p>Well done! We have covered a lot of ground in this chapter. I suggest digging out
some good music and chilling out for a bit.</p>
</div>
<div class="section" id="key-concepts">
<h2>Key concepts</h2>
<ul class="simple">
<li>Python is a powerful scripting language that is popular in the scientific community</li>
<li>You can explore Python’s syntax using its interactive mode</li>
<li>Variables and functions help us avoid having to repeat ourselves, the <a class="reference internal" href="glossary.html#term-dry"><span class="xref std std-term">DRY</span></a> principle</li>
<li>When naming variables and functions explicit trumps succinct</li>
<li>Loops are really powerful, they form the basis of automating repetitive tasks</li>
<li>Files are accessed using file handles</li>
<li>A file handle is a data structure that handles the book keeping of the
position within the file and the mode in which it was opened</li>
<li>The mode of the file handle determines how you will interact with the file</li>
<li>Read mode only allows reading of a file</li>
<li>Append mode will keep the existing content of a file and append to it</li>
<li>Write mode will delete any previous content before writing to it</li>
</ul>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2017, otu.<br/>
    </p>
  </div>
</footer>
  </body>
</html>